search-app/src/cmr/search/services/association_validation.clj

```diff
@@ -0,0 +1,260 @@
+(ns cmr.search.services.association-validation
+  "This contains functions for validating the business rules of an association between a collection
+   and a tag or a collection between a variable."
+  (:require
+   [cheshire.core :as json]
+   [clojure.set :as set]
+   [clojure.string :as string]
+   [cmr.common-app.services.search.query-execution :as qe]
+   [cmr.common-app.services.search.query-model :as cqm]
+   [cmr.common.services.errors :as errors]
+   [cmr.common.util :as util]
+   [cmr.search.services.tagging.json-schema-validation :as jv]
+   [cmr.search.services.tagging.tag-validation :as tv]
+   [cmr.search.services.messages.association-messages :as assoc-msg]
+   [cmr.transmit.metadata-db :as mdb]))
+
+(defn associations-json->associations
+  "Validates the associations json and returns the parsed json"
+  [associations-json]
+  (jv/validate-associations-json (tv/sanitized-json associations-json))
+  (->> (json/parse-string associations-json true)
+       (map #(set/rename-keys % {:concept_id :concept-id :revision_id :revision-id}))))
+
+(defn- get-inaccessible-concept-ids
+  "Returns the collection concept-ids within the given list that are invalid,
+  i.e. the collections for the given concept-ids do not exist or are not viewable by the token."
+  [context coll-concept-ids]
+  (when (seq coll-concept-ids)
+    (let [query (cqm/query {:concept-type :collection
+                            :condition (cqm/string-conditions :concept-id coll-concept-ids true)
+                            :page-size :unlimited
+                            :result-format :query-specified
+                            :result-fields [:concept-id]
+                            :skip-acls? false})
+          concept-ids (->> (qe/execute-query context query)
+                           :items
+                           (map :concept-id))]
+      (set/difference (set coll-concept-ids) (set concept-ids)))))
+
+(defn- get-bad-collection-revisions
+  "Returns the bad collection revisions of the given tag associations partitioned into a set of
```

Comment:
Remove the word "tag" from all of the docstrings that apply to both tag and variable associations. There are a bunch.